README.md
# Sque

Sque is painless interface to Amazons SQS, built for [node.js](http://nodejs.org).

## Status

**Current features:**

- 0.0.2 - SendMessage
- 0.0.2 - SendMessageBatch
- 0.0.1 - GetQueueAttributes
- 0.0.1 - ReceiveMessage
- 0.0.1 - DeleteMessage

**Feature roadmap:**

1. ListQueues / GetQueueUrl / CreateQueue / DeleteQueue / SetQueueAttributes
2. DeleteMessageBatch / ChangeMessageVisibility / ChangeMessageVisibilityBatch
3. AddPermission / RemovePermission

## Installation

    $ npm install sque

## Usage

```js
var Sque = require('sque');

// work with a specific queue:
var queue = Sque.getQueue({
  key      : '<AWS Key>',
  secret   : '<AWS Secret>',
  url      : '<AWS SQS queue url>'
});

/**
 * Queue attributes
 */

// get all queue attributes
queue.getAttributes(function( err, attributes ) {
  if ( err ) return console.log( err );
  console.log( attributes );
});

// get specific queue attribute
queue.getAttributes(['ApproximateNumberOfMessages'], function( err, attribute ) {
  if ( err ) return console.log( err );
  console.log( attribute );
});

// get multiple specific queue attributes
queue.getAttributes(['VisibilityTimeout', 'ApproximateNumberOfMessagesNotVisible'], function( err, attributes ) {
  if ( err ) return console.log( err );
  console.log( attributes );
});

/**
 * Queue subscription
 */

queue.subscribe({
  interval : 5000 // 5s polling interval; defaults to 10s if not sets
});

var messageCallback = function( message ) {
  console.log( 'queue message:' );
  console.dir( message );
  // ... do stuff with the message ...
  queue.deleteMessage( message.ReceiptHandle, function( err, success ) {
    if ( err ) return console.log( err );
    console.log( 'Hooray!' );
    console.dir( success );
  });
};
queue.on( 'message', messageCallback );

var messageErrorCallback = function( err ) {
  console.log( 'queue message error: ' );
  console.dir( err );
};
queue.on( 'message error', messageErrorCallback );

// unsubscribe
setTimeout(function() {
  queue.unsubscribe();
  queue.removeListener( 'message', messageCallback );
  queue.removeListener( 'message error', messageErrorCallback );
},20000);
```

## License

(The MIT License)

Copyright (c) 2012 rocket eleven GmbH <dominik@rocketeleven.de>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

HISTORY.md
0.0.2 / 2012-10-04
==================

  * SendMessage added
  * SendMessages added
  * Basic tests added

0.0.1 / 2012-10-02
==================

  * Initial release

test/sue.test.js
// module dependencies
var sque  = require('../lib/sque');
var nconf = require('nconf');

// load config file
nconf.file({ file: './test/config.json' });

// global var
var queue;

describe('sque', function () {

  describe('getQueue()', function() {
    it('should init without errors', function( done ) {
      queue = sque.getQueue({
        key      : nconf.get('AWS:Key'),
        secret   : nconf.get('AWS:Secret'),
        url      : nconf.get('Queue')
      });
      done();
    });
  });

  describe('getAttributes()', function() {
    it('should finish without errors', function( done ) {
      queue.getAttributes( done );
    });
  });

  describe('getMessage()', function() {
    it('should finish without errors', function( done ) {
      queue.getMessage( done );
    });
  });

  describe('sendMessage()', function() {
    it('should save without errors', function( done ) {
      queue.sendMessage({ foo : 'bar' }, done);
    });
  });

  describe('sendMessages()', function() {
    it('should save without errors', function( done ) {
      queue.sendMessages([{ foo : 'bar' }, { bar : 'foo' }], done);
    });
  });

});

lib/sque.js
/*!
 * Sque
 * Copyright (c) 2012 rocket eleven GmbH <dominik@rocketeleven.de>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var EventEmitter = require('events').EventEmitter;
var util         = require('util');

var _            = require( 'underscore' )._;
var aws          = require( 'aws-lib' );

var utils        = require( './utils' );

/**
 * Library version.
 */

exports.version = '0.0.2';

/**
 * Initialize a `Queue` with the given `url`.
 *
 * Required:
 *  - `key`     Amazon API key
 *  - `secret`  Amazon API secret
 *  - `url`     Queue url
 *
 * @param {string} url
 * @api public
 */

var Queue = function( options ) {
  if ( !options.key )    throw new Error( 'Amazon "API key" required' );
  if ( !options.secret ) throw new Error( 'Amazon "API secret" required' );
  if ( !options.url )    throw new Error( 'Amazon "Queue url" required' );

  EventEmitter.call(this);

  this.subscriptionActive = false;

  this.key      = options.key;
  this.secret   = options.secret;
  this.endpoint = utils.getEndpoint( options.url );
  this.path     = utils.getPath( options.url );
  this.client   = aws.createSQSClient( this.key, this.secret, {
    host : this.endpoint,
    path : this.path,
    version : '2011-10-01'
  });

};
util.inherits(Queue, EventEmitter);

/**
 * Returns one or all `attributes` of a queue with callback `fn( err, attributes )`.
 *
 * @param {Object} options
 * @param {Function} callback
 * @return {Function} fn
 * @api public
 */

Queue.prototype.getAttributes = function( options, fn ) {
  var self = this;
  var reqParams = {};
  if ( 'undefined' == typeof fn ) {
    fn = options;
    options = [];
    reqParams['AttributeName.1'] = 'All';
  }
  else {
    for ( var i = options.length - 1; i >= 0; i-- ) {
      var key = 'AttributeName.' + ( i + 1 );
      var value = options[i];
      reqParams[key] = value;
    }
  }
  self.client.call( 'GetQueueAttributes', reqParams, function( err, attributes ) {
    if ( err ) return fn( err );
    attributes = attributes['GetQueueAttributesResult']['Attribute'];
    if ( 'undefined' == typeof attributes ) return fn( null, {} );
    return fn( null, utils.refactorQueueAttributes(attributes) );
  });
};

/**
 * Retrieves one `message` from the queue with callback `fn( err, message )`.
 *
 * @param {Function} callback
 * @return {Function} fn
 * @api private
 */

Queue.prototype.getMessage = function( fn ) {
  var self = this;
  var reqOptions = {
    MaxNumberOfMessages : 1
  };
  self.client.call( 'ReceiveMessage', reqOptions, function( err, _results ) {
    if ( err ) return fn( err );
    // return if empty
    if ( _.isEmpty(_results['ReceiveMessageResult']['Message']) ) {
      return fn( null, [] );
    }
    var results = [];
    // check types
    if ( _.isArray(_results['ReceiveMessageResult']['Message']) ) {
      results = _results['ReceiveMessageResult']['Message'];
    }
    else if ( _.isObject(_results['ReceiveMessageResult']['Message']) ) {
      results.push( _results['ReceiveMessageResult']['Message'] );
    }
    // return results
    return fn( null, results );
  });
};

/**
 * Deletes a messages from the queue.
 *
 * @param {String} receiptHandle
 * @param {Function} callback
 * @return {Function} fn
 * @api public
 */

Queue.prototype.deleteMessage = function( receiptHandle, fn ) {
  var self = this;
  if ( !receiptHandle ) {
    return fn( new Error('"ReceiptHandle" required') );
  }
  var reqOptions = {
    ReceiptHandle : receiptHandle
  };
  self.client.call( 'DeleteMessage', reqOptions, function( err, _results ) {
    if ( err ) return fn( err );
    var results = _results;
    // return results
    return fn( null, results );
  });
};

/**
 * Send a messages to the queue.
 *
 * @param {Object} message
 * @param {Function} callback
 * @return {Function} fn
 * @api public
 */

Queue.prototype.sendMessage = function( message, fn ) {
  var self = this;
  
  if ( _.isEmpty(message) || 'function' == typeof message ) {
    return fn( new Error('"Message" required') );
  }

  var reqOptions = {
    MessageBody : JSON.stringify( message )
  };

  self.client.call( 'SendMessage', reqOptions, function( err, _results ) {
    if ( err ) return fn( err );
    var results = _results['SendMessageResult'];
    // return results
    return fn( null, results );
  });

};

/**
 * Send messagess to the queue.
 *
 * @param {Object} messages
 * @param {Function} callback
 * @return {Function} fn
 * @api public
 */

Queue.prototype.sendMessages = function( messages, fn ) {
  var self = this;
  
  if ( _.isEmpty(messages) || 'function' == typeof messages ) {
    return fn( new Error('"Messages" required') );
  }

  var reqOptions = {
  };

  for ( var i = 0; i < messages.length; i++ ) {
    reqOptions['SendMessageBatchRequestEntry.' + ( i + 1 ) +'.Id'] = utils.generateId(),
    reqOptions['SendMessageBatchRequestEntry.' + ( i + 1 ) +'.MessageBody'] = JSON.stringify( messages[i] );
  }

  self.client.call( 'SendMessageBatch', reqOptions, function( err, _results ) {
    if ( err ) return fn( err );
    var results = _results['SendMessageBatchResult']['SendMessageBatchResultEntry'];
    // return results
    return fn( null, results );
  });

};

/**
 * Starts the queue to constantly request `messages` with a given `interval`
 *
 * Optional:
 *   - `interval` Polling interval in ms; default : 10000
 *
 * @param {Object} options
 * @api public
 */

Queue.prototype.subscribe = function( options ) {
  var self = this;
  var interval = (options.interval) ? options.interval : 10000;
  var fetch = function () {
    self.getMessage( function( err, message ) {
      if ( err ) {
        self.emit('message error', { code : 500, msg : 'Failed to retrieve message.' });
      }
      else {
        if ( !_.isEmpty(message) ) self.emit('message', message[0] );
      }
      if ( self.subscriptionActive ) setTimeout( fetch, interval );
    });
  };
  fetch();
  self.subscriptionActive = true;
};

/**
 * Stops the queue from requesting `messages`
 *
 * @api public
 */

Queue.prototype.unsubscribe = function() {
  var self = this;
  self.subscriptionActive = false;
};

/**
 * Shortcut for `new Queue()`.
 *
 * @param {Object} options
 * @see Queue()
 * @api public
 */

exports.getQueue = function( options ) {
  return new Queue( options );
};

lib/utils.js
/*!
 * Sque
 * Copyright (c) 2012 rocket eleven GmbH <dominik@rocketeleven.de>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var _ = require( 'underscore' )._;

/**
 * Regex
 */

var parse_url = /^(?:([A-Za-z]+):)?(\/{0,3})([0-9.\-A-Za-z]+)(?::(\d+))?(?:\/([^?#]*))?(?:\?([^#]*))?(?:#(.*))?$/;

/**
 * Returns the AWS `endpoint` of a given `string`
 *
 * @param {string} string
 * @return {string} url
 * @api private
 */

exports.getEndpoint = function( string ) {
  if ( !string ) throw new Error( '"String" required' );
  var parts = parse_url.exec( string ); // gives us: url, scheme, slash, host, port, path, query, hash
  return parts[3];
};

/**
 * Returns the `path` of a given `string`
 *
 * @param {string} string
 * @return {string} path
 * @api private
 */

exports.getPath = function( string ) {
  if ( !string ) throw new Error( '"String" required' );
  var parts = parse_url.exec( string ); // gives us: url, scheme, slash, host, port, path, query, hash
  return '/' + parts[5];
};

/**
 * Returns a normalized `results` object of a given `attributes` object
 *
 * @param {Object} attributes
 * @return {Object} results
 * @api private
 */

exports.refactorQueueAttributes = function( attributes ) {
  var results = {};
  if ( !_.isArray(attributes) ) {
    results[attributes['Name']] = attributes['Value'];
  }
  else {
    for ( var i = 0; i < attributes.length; i++ ) {
      var key = attributes[i]['Name'];
      var value = attributes[i]['Value'];
      results[key] = value;
    }
  }
  return results;
};

/**
 * Returns an id
 *
 * @return {String} id
 * @api private
 */

exports.generateId = function() {
  var id = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
      return v.toString(16);
  });
  return id;
};

examples/app.js
var Sque = require('../');

// work with a specific queue:
var queue = Sque.getQueue({
  key      : '<AWS Key>',
  secret   : '<AWS Secret>',
  url      : '<AWS SQS queue url>'
});

/**
 * Queue attributes
 */

// get all queue attributes
queue.getAttributes(function( err, attributes ) {
  if ( err ) return console.log( err );
  console.log( attributes );
});

// get specific queue attribute
queue.getAttributes(['ApproximateNumberOfMessages'], function( err, attribute ) {
  if ( err ) return console.log( err );
  console.log( attribute );
});

// get multiple specific queue attributes
queue.getAttributes(['VisibilityTimeout', 'ApproximateNumberOfMessagesNotVisible'], function( err, attributes ) {
  if ( err ) return console.log( err );
  console.log( attributes );
});

/**
 * Send message(s)
 */

// single message
queue.sendMessage({ foo : 'bar' }, function( err, messageInformation ) {
  if ( err ) return console.log( err );
  console.log( messageInformation );
});

// multiple messages
var messages = [
  { foo : 'bar' },
  { bar : 'foo' }
];
queue.sendMessages(messages, function( err, messageInformations ) {
  if ( err ) return console.log( err );
  console.dir( messageInformations );
});


/**
 * Queue subscription
 */

// subscribe
queue.subscribe({
  interval : 5000 // 5s polling interval; defaults to 10s if not sets
});

var messageCallback = function( message ) {
  console.log( 'queue message:' );
  console.dir( message );
  // ... do stuff with the message ...
  queue.deleteMessage( message.ReceiptHandle, function( err, success ) {
    if ( err ) return console.log( err );
    console.log( 'Hooray!' );
    console.dir( success );
  });
};
queue.on( 'message', messageCallback );

var messageErrorCallback = function( err ) {
  console.log( 'queue message error: ' );
  console.dir( err );
};
queue.on( 'message error', messageErrorCallback );

// unsubscribe
setTimeout(function() {
  queue.unsubscribe();
  queue.removeListener( 'message', messageCallback );
  queue.removeListener( 'message error', messageErrorCallback );
},20000);